### ISSUE TITLE
Authorization failed when doing introspection

### ISSUE THREAD

**Issue by oscarwest** [aspnet-contrib/AspNet.Security.OAuth.Extensions#73]
Hello,
We're trying to introspect tokens towards a custom STS OAuth token issuer but we're having troubles getting everything working.

Goal: Introspect a ref token coming in, save the results in HttpContext.Authentication (or so) to then be able to access the claims later. The API Just wants to confirm the introspection to be "active" then let the user do controller actions(passing both authentication and authorization without an actual user)

Problem:
Introspection is successfull "The introspection response was successfully extracted: ... "active": true".
When introspection returns "active": "false" with introspection was not authenticated. Failure message: Authentication failed because the authorization server rejected the access token.
. Code continues to run into the controllers, we never fail the auth context.

`
info: Microsoft.AspNetCore.Authorization.DefaultAuthorizationService[2]
      Authorization failed for user: (null).
Microsoft.AspNetCore.Authorization.DefaultAuthorizationService:Information: Authorization failed for user: (null).
info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
      Authorization failed for the request at filter 'Microsoft.AspNetCore.Mvc.Authorization.AuthorizeFilter'.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Authorization failed for the request at filter 'Microsoft.AspNetCore.Mvc.Authorization.AuthorizeFilter'.
info: Microsoft.AspNetCore.Mvc.ChallengeResult[1]
      Executing ChallengeResult with authentication schemes (introspection).
Microsoft.AspNetCore.Mvc.ChallengeResult:Information: Executing ChallengeResult with authentication schemes (introspection).
info: AspNet.Security.OAuth.Introspection.OAuthIntrospectionMiddleware[12]
      AuthenticationScheme: introspection was challenged.
AspNet.Security.OAuth.Introspection.OAuthIntrospectionMiddleware:Information: AuthenticationScheme: introspection was challenged.
info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[2]
      Executed action Project.Olivia.Api.Controllers.MyController.GetAccountTransactionsAsync (Project.Olivia.Api) in 8869.5683ms
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executed action Project.Olivia.Api.Controllers.MyController.GetAccountTransactionsAsync (Project.Olivia.Api) in 8869.5683ms
Loaded '/Users/west/.nuget/packages/System.Runtime.CompilerServices.Unsafe/4.3.0/lib/netstandard1.0/System.Runtime.CompilerServices.Unsafe.dll'. Module was built without symbols.
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
      Request finished in 9288.3944ms 401 
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 9288.3944ms 401 
`

Setup:
ConfigureServices

```
services.AddAuthorization(options => 
{
    // Just passes context.Succeed(requirement); return Task.FromResult(0);
    options.AddPolicy("test", policy => policy.Requirements.Add(new AuthorizationRequirement()));
});

services.AddAuthentication();

```

Configure

```
app.UseOAuthIntrospection(options =>
{
    options.Authority = new Uri("*");
    options.Audiences.Add("*");
    options.AutomaticAuthenticate = false;
    options.AutomaticChallenge = false;
    options.SaveToken = true;
    options.ClientId = "*";
    options.ClientSecret = "*";
    options.RequireHttpsMetadata = false;
    options.MetadataAddress = new Uri("*/oauth/v2/metadata/.well-known/openid-configuration");
    options.AuthenticationScheme = "introspection";
    options.IncludeErrorDetails = true;
    options.Events = new OAuthIntrospectionEvents()
    {
        OnCreateTicket = context => {
            // context.HttpContext.User.AddIdentity( context.Identity );
            // context.HttpContext.Authentication.AuthenticateAsync("introspection");
            
            return Task.FromResult(0);
        },
        OnRetrieveToken = context => {
            return Task.FromResult(0);
        },
        OnValidateToken = context => {
            return Task.FromResult(0);
        },
        OnSendIntrospectionRequest = context => {
            return Task.FromResult(0);
        }
    };
});

```

Controller
```

[Authorize(Policy="test")]
[Authorize(ActiveAuthenticationSchemes = "introspection")]


```

**Comment by kevinchalet** [comment]
Please change the default log level to `Trace` (verbose) and post your entire logs file.

**Comment by oscarwest** [comment]
```
dbug: Microsoft.AspNetCore.Routing.Tree.TreeRouter[1]
      Request successfully matched the route with name '(null)' and template 'api/v1/theproject/account/transactions'.
dbug: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
      Executing action project.Olivia.Api.Controllers.theprojectController.GetAccountTransactionsAsync (project.Olivia.Api)
Loaded '/Users/west/.nuget/packages/Microsoft.IdentityModel.Tokens/5.1.3/lib/netstandard1.4/Microsoft.IdentityModel.Tokens.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Loaded '/usr/local/share/dotnet/shared/Microsoft.NETCore.App/1.1.2/System.Security.Cryptography.Encoding.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
The thread 1836527 has exited with code 0 (0x0).
info: AspNet.Security.OAuth.Introspection.OAuthIntrospectionMiddleware[0]
      The introspection response was successfully extracted: "theproject_token": "*", "sub": "theproject", "purpose": "access_token", "iss": "https://login.project.se", "active": true, "token_type": "bearer", "client_id": "theproject", "aud": "https://commonservices-test.projectbank.se", "nbf": 1497514540.0, "scope": "", "expired_scope": [], "exp": 1497517540.0, "iat": 1497516016.0, "delegationId": "e1c46f55-f285-4f3c-82e3-02235554fc93".
AspNet.Security.OAuth.Introspection.OAuthIntrospectionMiddleware:Information: The introspection response was successfully extracted: "theproject_token": "*", "sub": "theproject", "purpose": "access_token", "iss": "https://login.project.se", "active": true, "token_type": "bearer", "client_id": "theproject", "aud": "https://commonservices-test.projectbank.se", "nbf": 1497514540.0, "scope": "", "expired_scope": [], "exp": 1497517540.0, "iat": 1497516016.0, "delegationId": "e1c46f55-f285-4f3c-82e3-02235554fc93".
The thread 1836532 has exited with code 0 (0x0).
Loaded '/Users/west/.nuget/packages/Microsoft.AspNetCore.WebUtilities/1.1.2/lib/netstandard1.3/Microsoft.AspNetCore.WebUtilities.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Loaded '/usr/local/share/dotnet/shared/Microsoft.NETCore.App/1.1.2/System.Security.Cryptography.Primitives.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
dbug: Microsoft.AspNetCore.DataProtection.Repositories.FileSystemXmlRepository[37]

dbug: Microsoft.AspNetCore.DataProtection.Repositories.FileSystemXmlRepository[37]
      Reading data from file '/Users/west/.aspnet/DataProtection-Keys/key-c12de410-71c7-43b9-a6f2-8371b391e7e7.xml'.
dbug: Microsoft.AspNetCore.DataProtection.Repositories.FileSystemXmlRepository[37]
      Reading data from file '/Users/west/.aspnet/DataProtection-Keys/key-e4c30714-5582-456c-b763-3485e9bccb86.xml'.
dbug: Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager[18]
      Found key {c12de410-71c7-43b9-a6f2-8371b391e7e7}.
dbug: Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager[18]
      Found key {e4c30714-5582-456c-b763-3485e9bccb86}.
dbug: Microsoft.AspNetCore.DataProtection.KeyManagement.DefaultKeyResolver[13]
      Considering key {c12de410-71c7-43b9-a6f2-8371b391e7e7} with expiration date 2017-09-10 13:36:03Z as default key.
Loaded '/usr/local/share/dotnet/shared/Microsoft.NETCore.App/1.1.2/System.Security.Cryptography.Algorithms.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
dbug: Microsoft.AspNetCore.DataProtection.AuthenticatedEncryption.ConfigurationModel.ManagedAuthenticatedEncryptorDescriptor[11]
      Using managed symmetric algorithm 'System.Security.Cryptography.Aes'.
dbug: Microsoft.AspNetCore.DataProtection.AuthenticatedEncryption.ConfigurationModel.ManagedAuthenticatedEncryptorDescriptor[10]
      Using managed keyed hash algorithm 'System.Security.Cryptography.HMACSHA256'.
dbug: Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider[2]
      Using key {c12de410-71c7-43b9-a6f2-8371b391e7e7} as the default key.
trce: Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingBasedDataProtector[31]
      Performing protect operation to key {c12de410-71c7-43b9-a6f2-8371b391e7e7} with purposes ('/Users/west/projects/project-olivia/project.Olivia.Api', 'OAuthIntrospectionHandler', 'AccessTokenFormat', 'introspection').
info: Microsoft.AspNetCore.Authorization.DefaultAuthorizationService[2]
      Authorization failed for user: (null).
Microsoft.AspNetCore.Authorization.DefaultAuthorizationService:Information: Authorization failed for user: (null).
info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
      Authorization failed for the request at filter 'Microsoft.AspNetCore.Mvc.Authorization.AuthorizeFilter'.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Authorization failed for the request at filter 'Microsoft.AspNetCore.Mvc.Authorization.AuthorizeFilter'.
info: Microsoft.AspNetCore.Mvc.ChallengeResult[1]
      Executing ChallengeResult with authentication schemes (introspection).

```

**Comment by kevinchalet** [comment]
Looks like ASP.NET Core doesn't log failed authentications, which would explain why your dump doesn't contain the traces I'm looking for: https://github.com/aspnet/Security/blob/rel/1.1.1/src/Microsoft.AspNetCore.Authentication/AuthenticationHandler.cs#L223. I'll open a ticket on the right repo to track that.

The introspection middleware also returns the error code/description as part of the WWW-Authenticate header. Could you place try sending a request to your API with a valid token and share the HTTP response?

**Comment by kevinchalet** [comment]
Alternatively, enable automatic authentication. The error is logged by ASP.NET Core in this case: https://github.com/aspnet/Security/blob/1.0.0/src/Microsoft.AspNetCore.Authentication/AuthenticationHandler.cs#L107

**Comment by oscarwest** [comment]
Enabling automatic authentication and disabling all policies and removing [Authorize(ActiveAuthenticationSchemes = "introspection")] from the controller
`
options.AutomaticAuthenticate = true;
`

A valid token (active=true) fails like this:
`
The introspection response was successfully extracted:
trce: Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingBasedDataProtector[31]
      Performing protect operation to key {c12de410-71c7-43b9-a6f2-8371b391e7e7} with purposes ('/Users/west/projects/project-olivia/project.Olivia.Api', 'OAuthIntrospectionHandler', 'AccessTokenFormat', 'Bearer').
info: AspNet.Security.OAuth.Introspection.OAuthIntrospectionMiddleware[7]
      Bearer was not authenticated. Failure message: Authentication failed because the access token was not valid for this resource server.
AspNet.Security.OAuth.Introspection.OAuthIntrospectionMiddleware:Information: Bearer was not authenticated. Failure message: Authentication failed because the access token was not valid for this resource server.
`


And still falls through to the controller where calling this:
`
HttpContext.Authentication.GetAuthenticateInfoAsync("introspection")
`
Presents this exception:
"One or more errors occurred. (No authentication handler is configured to authenticate for the scheme: introspection)"
`
"   at Microsoft.AspNetCore.Http.Authentication.Internal.DefaultAuthenticationManager.<GetAuthenticateInfoAsync>d__12.MoveNext()"
`

Are we missing something here? Do we need to write our own authentication handler?
If the introspection returns active=false we want to short circuit the request with 401, otherwise we want to save the introspection response payload in HttpContext so we can access it inside the controllers. How would we do this? A nudge in the right direction would mean a lot.

/ Oscar

**Comment by kevinchalet** [comment]
The `Authentication failed because the access token was not valid for this resource server.` message indicates that the `aud` claim returned by the remote server doesn't match one of the audiences registered in the options.

Concerning the second error (thrown by ASP.NET Core), I suppose you're trying to use the introspection middleware before it's registered. Make sure your middleware are in the correct order.

**Comment by kevinchalet** [comment]
Closing, as I presume your issue should now be fixed.

**Comment by oscarwest** [comment]
@PinpointTownes We are still having problems setting this middleware up to work in a real world scenario.

When the OAuthIntrospectionHandler skips on the result, how would we intercept the failure so we can return 401 explicitely? (ex https://github.com/aspnet-contrib/AspNet.Security.OAuth.Extensions/blob/a120c105058f167ac5e72513c9242029b32b0a46/src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs#L67)

Also, when the Handler fails the request, how would be intercept this?
(ex https://github.com/aspnet-contrib/AspNet.Security.OAuth.Extensions/blob/a120c105058f167ac5e72513c9242029b32b0a46/src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs#L111)

Even when sending an active token the ticket and ClaimsPrincipal doesn't seem to get set and we get this after the OnCreateTicket has returned:
Authorization failed for user: (null).

Setting AutomaticAuthenticate = true, we get 
Authentication was skipped because no bearer token was received.
But the request continues since the auth is skipped.

Basically, is there an event we can listen to for passed/failed AuthenticateResult or by some other means short-circuit the request pipeline if Authentication fails? If not, do we need to write an AuthorizationHandler to check if the Authentication passed, and if so, copy over the Claims to create an Identity? By reading the source code it looks like this should already be done by the middleware.

**Comment by kevinchalet** [comment]
Since your questions are not specific to the introspection middleware - that uses the same events pattern as all the ASP.NET Core security middleware, I'd recommend opening a new ticket here: https://github.com/aspnet/Security/issues.

Alternatively, I can provide your company with consulting services, if you need dedicated help.

**Comment by oscarwest** [comment]
Thank you for your help 👍
